src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -49,24 +49,50 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
             anyfail = true;
         }
         MPIDI_XPMEM_global.segmaps[i].apid = -1;        /* get apid at the first communication  */
-
-        /* Init AVL tree based segment cache */
-        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache);
     }
+    MPIDU_Init_shm_barrier();
 
     /* Check to make sure all processes initialized XPMEM correctly. */
     if (anyfail) {
-        for (i = 0; i < num_local; i++) {
-            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache);
-        }
         /* Not setting an mpi_errno value here because we can handle the failure of XPMEM
          * gracefully. */
         goto fn_fail;
     }
 
     /* Initialize other global parameters */
+    MPIDI_XPMEM_global.num_pending_cnt = 0;
     MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);
 
+    /* Initialize coop counter */
+    MPIR_CHKPMEM_MALLOC(MPIDI_XPMEM_global.coop_counter_direct, MPIDI_XPMEM_cnt_t **,
+                        sizeof(MPIDI_XPMEM_cnt_t *) * num_local,
+                        mpi_errno, "xpmem direct counter array", MPL_MEM_SHM);
+    MPIR_CHKPMEM_MALLOC(MPIDI_XPMEM_global.coop_counter_seg_direct, MPIDI_XPMEM_seg_t **,
+                        sizeof(MPIDI_XPMEM_seg_t *) * num_local,
+                        mpi_errno, "xpmem direct counter segs", MPL_MEM_SHM);
+
+    uint64_t cnt_mem_addr = (uint64_t) (&MPIDI_XPMEM_cnt_mem_direct);
+    uint64_t remote_cnt_mem_addr;
+
+    MPIDU_Init_shm_put(&cnt_mem_addr, sizeof(uint64_t));
+    MPIDU_Init_shm_barrier();
+    for (i = 0; i < num_local; ++i) {
+        /* Init AVL tree based segment cache */
+        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache);
+        MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache_cnt);
```

Comment:
`MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache_ubuf);` is not a part of coop counter. Can you please init it in a separate loop? Besides, should we move `MPIDI_XPMEM_global.num_pending_cnt = 0;` into the scope of coop copy initialization?

Reply:
I don't think add another loop to specifically initialize `segcache_ubuf` is good. One loop is clear and obvious. I can add extra comments to explain what we are initializing for. I moved `MPIDI_XPMEM_global.num_pending_cnt = 0;` to cooperative initialization scoop